{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b7764602",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# 1000 Genomes analysis on BigQuery"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da31a4cc",
   "metadata": {},
   "source": [
    "### 1. Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb90f658",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "Authors: Prathima Vembu <br>\n",
    "Notebook kernel: R (https://irkernel.github.io/installation/)<br>\n",
    "Recommended Google Cloud Platform machine type: n1-standard-1 <br>\n",
    "Predicted runtime: ~30 minutes <br>\n",
    "Predicted cost for analysis <br>\n",
    "1. Variant optimized table : 83.40 USD per one month (~100 USD including allele frequency computation) <br>\n",
    "2. Sample optimized table : 0.00 USD per one month first \n",
    "*Note: 1 TB of query data processed per month is free* <br>\n",
    "\n",
    "This notebook includes methods for reproducing several figures in the 1000 Genomes phase 3 publication (https://doi.org/10.1038/nature15393)<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f056c5ee",
   "metadata": {},
   "source": [
    "##### 1.1: Comparing variant optimized table with sample optimized table\n",
    "||Variant optimized|Sample optimized|\n",
    "|---|---|---|\n",
    "|User specified |No, generated by default|Yes, use ```--sample_lookup_optimized_output_table``` parameter|\n",
    "|Number of tables created |25 (23 chromosome, sample_info, residual_info)|25 (23 chromosome, sample_info, residual_info)|\n",
    "|Analysis |Variant lookup query - Find all variants within a enomic region|Sample lookup query - Find all variants in Sample 'Z'|\n",
    "|Storage cost|No additional storage cost|Additional storage cost (for 25 tables), recovered when running queries for sample lookup analysis|\n",
    "|Call field|Call field *not* unnested |Call field unnested|\n",
    "|New field|No new column added|New column ```sample_id``` added which has the same value as ```call.sample_id``` column|\n",
    "|1KG table size|1000_genomes_phase_3_variants_20150220 - 5.38 TB|1000_genomes_phase_3_optimized_schema_variants_20150220 - 1.94 TB|"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da46a038",
   "metadata": {},
   "source": [
    "##### 1.2: Table location \n",
    "Variant optimized table : ```bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220``` <br>\n",
    "Sample optimized table : ```bigquery-public-data.human_genome_variants.1000_genomes_phase_3_optimized_schema_variants_20150220```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "065025bd",
   "metadata": {},
   "source": [
    "##### 1.3: Query cost comparison \n",
    "\n",
    "| Analysis | Variant table(1000Genomes variant table)| Data processed | Sample table(1000Genomes sample optimized table) | Data processed |\n",
    "|---|---|---|---|---|\n",
    "|Distinct sample_id‚Äôs|3.70 USD |1.74 TB | 0.46 USD|0.092 TB|\n",
    "|Number of mapped and missing ID‚Äôs|3.70 USD|1.74 TB|0.46 USD|0.092 TB|\n",
    "|Number of variant sites per sample| 19.00 USD|4.8 TB|1.27 USD|0.255 TB|\n",
    "|Average singletons|19.00 USD|4.8 TB|1.28 USD|0.256 TB|\n",
    "|Average singletons and rare variants |19.00 USD|4.8 TB|1.28 USD|0.256 TB|\n",
    "|Distribution of Allele Frequency|0.00 USD|0.001 TB|0.00 USD|0.001 TB|\n",
    "|Z-score computation|19.00 USD|4.8 TB|1.27 USD|0.255 TB|\n",
    "|Total|83.4 USD|22.68 TB|6.02 USD|1.207 TB|\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddd3be80",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### 2. Installing necessary packages in R\n",
    "Before staring analysis, install and load the below packages in R. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32710832",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#helps connecting R to DBMS\n",
    "install.packages(\"DBI\") \n",
    "#helps to work with data on cloud storage\n",
    "install.packages(\"bigrquery\") \n",
    "#package to perform multiple data manipulation functions\n",
    "install.packages(\"pillar\") \n",
    "#package to restructure and aggregate dataframes\n",
    "install.packages(\"reshape2\") \n",
    "#package for creating graphics \n",
    "install.packages(\"ggplot2\") \n",
    "\n",
    "library(DBI)\n",
    "library(bigrquery)\n",
    "library(pillar)\n",
    "library(reshape2)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cf04afc",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### 3. Authenticate and connect to BigQuery \n",
    "Connect to the Google BigQuerydatabase account using ```dbconnect```. Replace ```BILLING_ACCOUNT``` with the name of the Google Cloud project you want charges to be billed to."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd87d095",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#dbconnect creates a connection to the DBMS\n",
    "gcp_account<-dbConnect(bigquery(),\"bigquery-public-data\", \n",
    "                       dataset=\"human_genome_variants\", \n",
    "                       billing = \"BILLING_ACCOUNT\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc21eb2f",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### 4. Analysis\n",
    "\n",
    "##### 4.1: Check to confirm the number of samples present in the BigQuery table (from the paper, n=2504). <br>\n",
    "Two tables are available for the 1000 Genomes project which can be found under *bigquery-public-data.human_genome_variants* <br>\n",
    "1. 1000_genomes_phase_3_variants_20150220 : The Variant table<br>\n",
    "2. 1000_genomes_sample_info : The Sample table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27b9d41f",
   "metadata": {
    "scrolled": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Number of samples in the Variant table\n",
    "#Runtime - 20s\n",
    "#Data processed - 0.092 TB\n",
    "variant_sample_count_sql<-\"SELECT\n",
    "COUNT(distinct call.name) AS Sample_count\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v, UNNEST(v.call) AS call\"\n",
    "\n",
    "variant_sample_count<-dbGetQuery(gcp_account,variant_sample_count_sql)\n",
    "variant_sample_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c9bb8a7",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Number of samples in the Sample table \n",
    "#Runtime - 1.8s\n",
    "#Data processed - 30.76KB\n",
    "sample_table_count_sql<-\"SELECT\n",
    "COUNT(Sample) AS Sample_count\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`\"\n",
    "\n",
    "sample_table_count<-dbGetQuery(gcp_account, sample_table_count_sql)\n",
    "sample_table_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f9b2039",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "Given that both these tables have different number of distinct sample ID's, it is important to make sure that the list of sample ID's in the ***variant table*** is a subset of the ID's from the ***sample table*** to prevent data loss. <br>\n",
    "The query below computes the number of mapped ID's and number of missing ID's. All ID's from the ***variant table*** match with entries in the ***sample table***. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a77e377",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Compute the number of mapped ID's and the number of missing ID's\n",
    "#Runtime - 20s\n",
    "#Data processed - 1.74TB\n",
    "mapped_missing_values_sql<-\"WITH cte1 AS (\n",
    "SELECT DISTINCT call.name AS call_name\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v,UNNEST(v.call) AS call\n",
    "), cte2 AS (SELECT distinct Sample AS sample_name \n",
    " FROM `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`\n",
    "),cte3 AS(SELECT call_name, sample_name \n",
    "  FROM cte1 \n",
    "  RIGHT JOIN cte2 \n",
    "  ON cte1.call_name=cte2.sample_name\n",
    ")\n",
    "SELECT SUM(CASE WHEN call_name IS NULL THEN 1 ELSE 0 END) AS missing_values, \n",
    "SUM(CASE WHEN call_name IS NOT NULL THEN 1 ELSE 0 END) AS mapped_values FROM cte3\"\n",
    "\n",
    "mapped_missing_values<-dbGetQuery(gcp_account, mapped_missing_values_sql)\n",
    "mapped_missing_values\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25aa1138",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "##### 4.2: Number of variant sites per genome \n",
    "Refer: [Figure 1b](https://www.nature.com/articles/nature15393/figures/1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cc05f4d",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Compute the number of variant sites per genome\n",
    "#Runtime - 1.9s\n",
    "#Data processed - 4.8 TB\n",
    "\n",
    "variant_sites_per_genome_sql<-\"SELECT s.Population,s.Sample,COUNT(reference_name) as variant_count,\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v, UNNEST(v.call) AS call\n",
    "INNER JOIN `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`s\n",
    "ON call.name=s.Sample\n",
    "WHERE EXISTS (SELECT gt FROM UNNEST(call.genotype) gt WHERE gt > 0 AND reference_name NOT IN ('X','Y','MT'))\n",
    "GROUP BY 1,2 ORDER BY variant_count DESC\"\n",
    "\n",
    "variant_sites_per_genome<-dbGetQuery(gcp_account, variant_sites_per_genome_sql)\n",
    "head(variant_sites_per_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb75711f",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Runtime - 2.5s\n",
    "\n",
    "#Make plots wider on jupyter notebooks\n",
    "options(repr.plot.width=15, repr.plot.height=8)\n",
    "\n",
    "#Set the order of the labels \n",
    "variant_sites_per_genome$Population<-factor(variant_sites_per_genome$Population,levels = unique(variant_sites_per_genome$Population))\n",
    "\n",
    "#Generate the plot\n",
    "ggplot(variant_sites_per_genome, aes(y=variant_count,x=Population, color=Population))+\n",
    "geom_point()+theme(axis.text.x = element_blank(),axis.ticks.x = element_blank())+\n",
    "scale_y_continuous(breaks = c(3800000,4000000,4200000,4400000,4600000,4800000,5000000,5200000,5400000,5600000),\n",
    "                   minor_breaks = NULL, labels = function(l) {paste0(round(l/1e6,1),\"M\")})+\n",
    "labs(title = \"Variant sites per genome across populations\",x = \"Individual\",y = \"Variant sites\")+\n",
    "theme(plot.title = element_text(size=22),axis.text=element_text(size=14),axis.title=element_text(size=18))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77328be7",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "##### 4.3: The average number of singletons per genome\n",
    "Refer: [Figure 1c](https://www.nature.com/articles/nature15393/figures/1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fee7d7af",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Compute the number of singletons per genome\n",
    "#Runtime - 1.5s\n",
    "#Data processed - 4.8TB\n",
    "singletons_per_genome_sql<-\"WITH cte1 AS (SELECT s.Population,Sample,\n",
    "COUNT(reference_name) as variant_sites,COUNTIF(AC=1) AS Singletons,\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v,\n",
    "UNNEST(v.alternate_bases) AS Alternate, UNNEST(v.call) AS call\n",
    "INNER JOIN `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`s\n",
    "ON call.name=s.Sample\n",
    "WHERE EXISTS(SELECT gt FROM UNNEST(call.genotype) gt WHERE gt>0 ) AND reference_name NOT IN ('X','Y','MT')\n",
    "GROUP BY 1,2)\n",
    "SELECT Population, AVG(Singletons) AS Average_Singletons FROM cte1 GROUP BY 1 ORDER BY 2 DESC\"\n",
    "\n",
    "singletons_per_genome<-dbGetQuery(gcp_account, singletons_per_genome_sql)\n",
    "head(singletons_per_genome)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "531b37b4",
   "metadata": {},
   "source": [
    "The output table is sorted in a descending order by the ***Average_Singletons*** column. Output dataframe has 2 columns, the ***Population*** group and the ***Average_Singletons*** computed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aed2d7bd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Runtime - 0.9s\n",
    "\n",
    "#Make plots wider on jupyter notebooks\n",
    "options(repr.plot.width=15, repr.plot.height=8)\n",
    "\n",
    "#Setting the labels, AC=1\n",
    "singletons_per_genome$Population<-factor(singletons_per_genome$Population,levels = unique(singletons_per_genome$Population))\n",
    "\n",
    "#Generate the plot, AC=1\n",
    "ggplot(singletons_per_genome,aes(y=Average_Singletons,x=Population))+\n",
    "geom_bar(aes(fill=Population),stat = \"identity\",position=\"dodge\")+\n",
    "scale_y_continuous(breaks = c(0,2000,4000,6000,8000,10000,12000,14000,16000,18000,20000,22000),\n",
    "                   minor_breaks = NULL, labels = function(l) {paste0(round(l/1e3,1),\"K\")})+\n",
    "labs(title = \"Average number of singletons per genome\",x = \"Population\",y = \"Singletons per genome\")+\n",
    "theme(plot.title = element_text(size=22),axis.text=element_text(size=14),axis.title=element_text(size=18))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "358d47bd",
   "metadata": {},
   "source": [
    "##### 4.4: Distribution of allele frequency \n",
    "Refer: Extended Figure 3a\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d7a4a03",
   "metadata": {},
   "source": [
    "Allele frequency has been extracted from the 1KG table and split into 20 buckets. This data is then plotted as a bar graph to see th allele frequency distribution. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5bf815e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Bucketing the allele frequency\n",
    "#Runtime - 16s\n",
    "#Data processed - 650.58MB\n",
    "af_sql<-\"SELECT COUNT(AF) AF_Count, AF_bucket, \n",
    "FROM (SELECT  AF, CASE WHEN AF=0 THEN 0\n",
    "                       WHEN AF>0 AND AF<=0.05 THEN 1\n",
    "                       WHEN AF>0.05 AND AF<=0.1 THEN 2\n",
    "                       WHEN AF>0.1 AND AF<=0.15 THEN 3\n",
    "                       WHEN AF>0.15 AND AF<=0.2 THEN 4\n",
    "                       WHEN AF>0.2 AND AF<=0.25 THEN 5\n",
    "                       WHEN AF>0.25 AND AF<=0.3 THEN 6\n",
    "                       WHEN AF>0.3 AND AF<=0.35 THEN 7\n",
    "                       WHEN AF>0.35 AND AF<=0.4 THEN 8\n",
    "                       WHEN AF>0.4 AND AF<=0.45 THEN 9\n",
    "                       WHEN AF>0.45 AND AF<=0.5 THEN 10\n",
    "                       WHEN AF>0.5 AND AF<=0.55 THEN 11\n",
    "                       WHEN AF>0.55 AND AF<=0.6 THEN 12\n",
    "                       WHEN AF>0.6 AND AF<=0.65 THEN 13\n",
    "                       WHEN AF>0.65 AND AF<=0.7 THEN 14\n",
    "                       WHEN AF>0.7 AND AF<=0.75 THEN 15\n",
    "                       WHEN AF>0.75 AND AF<=0.8 THEN 16\n",
    "                       WHEN AF>0.8 AND AF<=0.85 THEN 17\n",
    "                       WHEN AF>0.85 AND AF<=0.9 THEN 18\n",
    "                       WHEN AF>0.9 AND AF<=0.95 THEN 19\n",
    "                       WHEN AF>0.95 AND AF<=1.0 THEN 20\n",
    "                       ELSE -1 END AS AF_bucket\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v,UNNEST(v.alternate_bases) AS Alternate )\n",
    "GROUP BY AF_bucket\n",
    "ORDER BY AF_bucket\"\n",
    "\n",
    "af<-dbGetQuery(gcp_account, af_sql)\n",
    "af"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "868c9583",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Runtime - 0.4s\n",
    "\n",
    "##REMOVING FIRST ROW - AF=0\n",
    "#Make plots wider on jupyter notebooks\n",
    "options(repr.plot.width=23, repr.plot.height=8)\n",
    "\n",
    "#Removing AF=0 row \n",
    "af_mod = af[-1,]\n",
    "\n",
    "#Generating the graph for variant tables \n",
    "ggplot(af_mod, aes(x=AF_bucket, y=AF_Count))+\n",
    "geom_bar(stat = \"identity\",fill=\"red\")+\n",
    "labs(title = \"Variant table - Allele Frequency distribution\",x = \"Allele frequency range\",y = \"Count(allele frequency)\")+\n",
    "theme(plot.title = element_text(size=22),axis.text=element_text(size=14),axis.title=element_text(size=18), \n",
    "      legend.title = element_blank(),legend.text = element_blank())+\n",
    "scale_x_continuous(breaks=c(1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20), minor_breaks=NULL,labels=c(\"0.05\",\"0.1\",\"0.15\", \"0.2\", \"0.25\", \"0.3\", \"0.35\", \"0.4\", \"0.45\", \n",
    "      \"0.5\", \"0.55\", \"0.6\", \"0.65\", \"0.7\", \"0.75\", \"0.8\", \"0.85\", \"0.9\", \"0.95\",\"1.0\"))+\n",
    "geom_text(aes(label = scales::comma(AF_Count)),vjust=-0.2,size=5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d50310a",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "##### 4.5 : Number of rare variants and singletons per genome <br>\n",
    "Refer: [Extended Figure 3b](https://www.nature.com/articles/nature15393/figures/7)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5dbbb80",
   "metadata": {
    "scrolled": true,
    "slideshow": {
     "slide_type": "fragment"
    },
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Compute the number of rare variants and singletons per genome\n",
    "#Runtime - 43s\n",
    "#Data processed - 4.8TB\n",
    "rare_var_and_singletons_per_genome_sql<-\"WITH cte1 AS (SELECT Population,Sample,\n",
    "COUNT(reference_name) as variant_sites,\n",
    "COUNTIF(AF<0.005) AS Rare_Variants,\n",
    "COUNTIF(AC=1) AS Singletons,\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v,\n",
    "UNNEST(v.alternate_bases) AS Alternate, UNNEST(v.call) AS call\n",
    "INNER JOIN `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`s\n",
    "ON call.name=s.Sample\n",
    "WHERE EXISTS(SELECT gt FROM UNNEST(call.genotype) gt WHERE gt>0 ) AND reference_name NOT IN ('X','Y','MT')\n",
    "GROUP BY 1,2) SELECT Population, AVG(Rare_variants) AS Average_rare_variants, AVG(Singletons) AS Average_singletons \n",
    "FROM cte1 GROUP BY 1 ORDER BY Average_rare_variants,Average_singletons DESC\"\n",
    "\n",
    "rare_var_and_singletons_per_genome<-dbGetQuery(gcp_account, rare_var_and_singletons_per_genome_sql)\n",
    "head(rare_var_and_singletons_per_genome)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "561aa8f7",
   "metadata": {},
   "source": [
    "The output table contains 3 colums, the ***Population***, ***Average_rare_variants*** computed and the ***Average_singletons*** computed. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c208842",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Runtime - 1.5s\n",
    "\n",
    "#Make plots wider on jupyter notebooks\n",
    "options(repr.plot.width=15, repr.plot.height=8)\n",
    "\n",
    "#Set the order of the labels \n",
    "rare_var_and_singletons_per_genome$Population<-factor(rare_var_and_singletons_per_genome$Population,levels = unique(rare_var_and_singletons_per_genome$Population))\n",
    "\n",
    "#Generating the graph \n",
    "ggplot(rare_var_and_singletons_per_genome,aes(x=Population, y=Average_rare_variants))+\n",
    "geom_bar(stat = \"identity\", position = \"stack\", fill=\"grey\")+\n",
    "geom_bar(data = subset(rare_var_and_singletons_per_genome,variable=Average_singletons),\n",
    "         aes(x=Population, y=Average_singletons, fill=Population),stat = \"identity\")+\n",
    "scale_y_continuous(breaks = c(0,20000,40000,60000,80000,100000,120000,140000,160000,180000,200000,220000,\n",
    "                              240000,260000,280000),labels = function(l) {paste0(round(l/1e3,1),\"K\")})+\n",
    "labs(title = \"Average number of rare variants (grey) with average number of singletons (coloured)\",\n",
    "     y = \"Rare variants per genome\")+\n",
    "theme(plot.title = element_text(size=20),axis.text=element_text(size=14),axis.title=element_text(size=18))\n"
   ]
  },
  {
   "attachments": {
    "Screen%20Shot%202022-04-12%20at%208.06.44%20AM.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAAawAAAC8CAYAAAAkaZqDAAAMamlDQ1BJQ0MgUHJvZmlsZQAASImVVwdYU8kWnluSkJDQAhGQEnoTRHqREkKLICBVsBGSQEKJMSGo2FFRwbWLKFZ0VUTRtQCyqIhdWQR7XywoKOuiLoqi8iYkoOu+8r3zfXPnv2fO/KdkJncGAM1erkSSjWoBkCPOlcaGBTHHJ6cwSc8BAuiAArQAhcuTSVgxMZEAymD/d3l/C1pDue6o4Prn+H8VHb5AxgMAmQhxGl/Gy4G4AQB8M08izQWAqNBbTM+VKPB8iHWlMECI1ylwhhLvVeA0Ja4fsImPZUPcAoAalcuVZgCg8QDqmXm8DMij8RliZzFfJAZAcwTE/jwhlw+xIvYROTlTFbgUYltoL4EYxgO80r7jzPgbf9oQP5ebMYSVeQ2IWrBIJsnmzvw/S/O/JSdbPujDGjaqUBoeq8gf1vBO1tQIBaZC3CVOi4pW1BriXhFfWXcAUIpQHp6gtEeNeDI2rB9gQOzM5wZHQGwEcag4OypSpU9LF4VyIIarBZ0hyuXEQ6wP8VKBLCROZbNdOjVW5QutTpeyWSr9Ra50wK/C1yN5VgJLxf9WKOCo+DGNfGF8EsQUiC3zRIlREGtA7CTLiotQ2YzOF7KjBm2k8lhF/JYQxwrEYUFKfiwvXRoaq7IvypEN5ottF4o4USp8OFcYH66sD3aWxx2IH+aCtQjErIRBHoFsfORgLnxBcIgyd6xDIE6IU/H0SnKDYpVzcYokO0Zlj5sLssMUenOI3WR5caq5eGIuXJxKfjxdkhsTr4wTz8/kjolRxoOvApGADYIBE8hhSwNTQSYQNXfVdME35Ugo4AIpyAAC4KjSDM5IGhgRw2ccyAd/QCQAsqF5QQOjApAH9V+GtMqnI0gfGM0bmJEFnkOcAyJANnyXD8wSD3lLBM+gRvQP71zYeDDebNgU4/9eP6j9pmFBTaRKIx/0yNQctCSGEIOJ4cRQoh1uiPvjvngkfAbC5oJ74d6DeXyzJzwntBKeEG4S2gh3p4gKpD9EORa0Qf5QVS3Svq8Fbg053fEg3A+yQ2acgRsCR9wN+mHhAdCzO9SyVXErqsL8gftvGXz3a6jsyM5klDyMHEi2/XGmhr2G+xCLotbf10cZa9pQvdlDIz/6Z39XfT7sI360xJZiR7AL2GnsElaP1QAmdgqrxZqwEwo8tLqeDayuQW+xA/FkQR7RP/xxVT4VlZQ5Vzp3On9WjuUKZuQqNh57qmSmVJQhzGWy4NdBwOSIeU4jmC7OLi4AKL41yr+vd4yBbwjCuPxNV/AQAL/k/v7++m+6SLh/j3bA7d/1TWdTCQDtJAAXF/Pk0jylDlc8CPBfQhPuNANgAiyALczHBXgAXxAIQsAYEA3iQTKYDKsshOtcCqaD2WABKATFYBVYDzaBbWAn2AsOgMOgBtSD0+A8uAJawE1wH66edvAKdIP3oA9BEBJCQ+iIAWKKWCEOiAvihfgjIUgkEoskI6lIBiJG5MhsZCFSjKxBNiE7kArkF+Q4chq5hLQid5HHSCfyFvmEYigV1UWNUWt0JOqFstAINB6dhGag09B8dBG6Ai1Fy9H9aDV6Gr2C3kTb0FdoDwYwdYyBmWGOmBfGxqKxFCwdk2JzsSKsBCvHqrA6+Dtfx9qwLuwjTsTpOBN3hCs4HE/Aefg0fC6+HN+E78Wr8bP4dfwx3o1/JdAIRgQHgg+BQxhPyCBMJxQSSgi7CccI5+Beaie8JxKJDKIN0RPuxWRiJnEWcTlxC/EgsYHYSnxK7CGRSAYkB5IfKZrEJeWSCkkbSftJp0jXSO2kXjV1NVM1F7VQtRQ1sVqBWonaPrWTatfUXqj1kbXIVmQfcjSZT55JXkneRa4jXyW3k/so2hQbih8lnpJJWUAppVRRzlEeUN6pq6ubq3urj1MXqc9XL1U/pH5R/bH6R6oO1Z7Kpk6kyqkrqHuoDdS71Hc0Gs2aFkhLoeXSVtAqaGdoj2i9GnQNJw2OBl9jnkaZRrXGNY3XmmRNK02W5mTNfM0SzSOaVzW7tMha1lpsLa7WXK0yreNat7V6tOnao7SjtXO0l2vv076k3aFD0rHWCdHh6yzS2alzRucpHaNb0Nl0Hn0hfRf9HL1dl6hro8vRzdQt1j2g26zbraej56aXqDdDr0zvhF4bA2NYMziMbMZKxmHGLcanYcbDWMMEw5YNqxp2bdgH/eH6gfoC/SL9g/o39T8ZMA1CDLIMVhvUGDw0xA3tDccZTjfcanjOsGu47nDf4bzhRcMPD79nhBrZG8UazTLaadRk1GNsYhxmLDHeaHzGuMuEYRJokmmyzuSkSacp3dTfVGS6zvSU6UumHpPFzGaWMs8yu82MzMLN5GY7zJrN+sxtzBPMC8wPmj+0oFh4WaRbrLNotOi2NLUcaznbstLynhXZystKaLXB6oLVB2sb6yTrJdY11h02+jYcm3ybSpsHtjTbANtptuW2N+yIdl52WXZb7FrsUXt3e6F9mf1VB9TBw0HksMWhdQRhhPcI8YjyEbcdqY4sxzzHSsfHTgynSKcCpxqn1yMtR6aMXD3ywsivzu7O2c67nO+P0hk1ZlTBqLpRb13sXXguZS43XGmuoa7zXGtd37g5uAnctrrdcae7j3Vf4t7o/sXD00PqUeXR6Wnpmeq52fO2l65XjNdyr4veBO8g73ne9d4ffTx8cn0O+/zp6+ib5bvPt2O0zWjB6F2jn/qZ+3H9dvi1+TP9U/23+7cFmAVwA8oDngRaBPIDdwe+YNmxMln7Wa+DnIOkQceCPrB92HPYDcFYcFhwUXBziE5IQsimkEeh5qEZoZWh3WHuYbPCGsIJ4RHhq8Nvc4w5PE4Fp3uM55g5Y85GUCPiIjZFPIm0j5RG1o1Fx44Zu3bsgyirKHFUTTSI5kSvjX4YYxMzLebXccRxMePKxj2PHRU7O/ZCHD1uSty+uPfxQfEr4+8n2CbIExoTNRMnJlYkfkgKTlqT1DZ+5Pg5468kGyaLkmtTSCmJKbtTeiaETFg/oX2i+8TCibcm2UyaMenSZMPJ2ZNPTNGcwp1yJJWQmpS6L/UzN5pbzu1J46RtTuvmsXkbeK/4gfx1/E6Bn2CN4EW6X/qa9I4Mv4y1GZ3CAGGJsEvEFm0SvckMz9yW+SErOmtPVn92UvbBHLWc1JzjYh1xlvjsVJOpM6a2ShwkhZK2aT7T1k/rlkZId8sQ2SRZba4uPNQ3yW3li+WP8/zzyvJ6pydOPzJDe4Z4RtNM+5nLZr7ID83/eRY+izercbbZ7AWzH89hzdkxF5mbNrdxnsW8RfPa54fN37uAsiBrwW8FzgVrCv5amLSwbpHxovmLni4OW1xZqFEoLby9xHfJtqX4UtHS5mWuyzYu+1rEL7pc7FxcUvx5OW/55Z9G/VT6U/+K9BXNKz1Wbl1FXCVedWt1wOq9a7TX5K95unbs2up1zHVF6/5aP2X9pRK3km0bKBvkG9pKI0trN1puXLXx8ybhpptlQWUHNxttXrb5wxb+lmtbA7dWbTPeVrzt03bR9js7wnZUl1uXl+wk7szb+XxX4q4LP3v9XLHbcHfx7i97xHva9sbuPVvhWVGxz2jfykq0Ul7ZuX/i/pYDwQdqqxyrdhxkHCw+BA7JD738JfWXW4cjDjce8TpSddTq6OZj9GNF1Uj1zOruGmFNW21ybevxMccb63zrjv3q9OueerP6shN6J1aepJxcdLL/VP6pngZJQ9fpjNNPG6c03j8z/syNs+PONp+LOHfxfOj5MxdYF05d9LtYf8nn0vHLXpdrrnhcqW5ybzr2m/tvx5o9mquvel6tbfFuqWsd3XryWsC109eDr5+/wblx5WbUzdZbCbfu3J54u+0O/07H3ey7b+7l3eu7P/8B4UHRQ62HJY+MHpX/bvf7wTaPthOPgx83PYl7cv8p7+mrZ7Jnn9sXPac9L3lh+qKiw6WjvjO0s+XlhJftrySv+roK/9D+Y/Nr29dH/wz8s6l7fHf7G+mb/rfL3xm82/OX21+NPTE9j97nvO/7UNRr0Lv3o9fHC5+SPr3om/6Z9Ln0i92Xuq8RXx/05/T3S7hS7sBRAIMNTU8H4O0eeE5IBoAO722UCcq74IAgyvvrAAL/CSvviwPiAUAV7BTHeHYDAIdgs54PueG74ggfHwhQV9ehphJZuquLkosKb0KE3v7+d8YAkOoA+CLt7+/b0t//ZRcM9i4ADdOUd1CFEOGdYbu/At3U588HP4jyfvpdjj/2QBGBG/ix/xcvLo/7+xEQVgAAAIplWElmTU0AKgAAAAgABAEaAAUAAAABAAAAPgEbAAUAAAABAAAARgEoAAMAAAABAAIAAIdpAAQAAAABAAAATgAAAAAAAACQAAAAAQAAAJAAAAABAAOShgAHAAAAEgAAAHigAgAEAAAAAQAAAaygAwAEAAAAAQAAALwAAAAAQVNDSUkAAABTY3JlZW5zaG90j5IEjAAAAAlwSFlzAAAWJQAAFiUBSVIk8AAAAdZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IlhNUCBDb3JlIDYuMC4wIj4KICAgPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4KICAgICAgPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIKICAgICAgICAgICAgeG1sbnM6ZXhpZj0iaHR0cDovL25zLmFkb2JlLmNvbS9leGlmLzEuMC8iPgogICAgICAgICA8ZXhpZjpQaXhlbFlEaW1lbnNpb24+MTg4PC9leGlmOlBpeGVsWURpbWVuc2lvbj4KICAgICAgICAgPGV4aWY6UGl4ZWxYRGltZW5zaW9uPjQyODwvZXhpZjpQaXhlbFhEaW1lbnNpb24+CiAgICAgICAgIDxleGlmOlVzZXJDb21tZW50PlNjcmVlbnNob3Q8L2V4aWY6VXNlckNvbW1lbnQ+CiAgICAgIDwvcmRmOkRlc2NyaXB0aW9uPgogICA8L3JkZjpSREY+CjwveDp4bXBtZXRhPgrWul+LAAAAHGlET1QAAAACAAAAAAAAAF4AAAAoAAAAXgAAAF4AABLRhK3K1QAAEp1JREFUeAHsnQeQFEUUht9hVlDArGAAE2YFRFEMKCgq5oAKKIoJjKWWWaTMSjAnCqUUAyZURFRERcWEopgVMWAGxBxR2/d32VMzO3PL7N7u3vbs31VXM9Pb3dPzvbl+PR3eqzMahIEESIAESIAEqpxAHRVWlUuI1SMBEiABErAEqLD4IpAACZAACXhBgArLCzGxkiRAAiRAAlRYfAdIgARIgAS8IECF5YWYWEkSIAESIAEqLL4DJEACJEACXhCgwvJCTKwkCZAACZAAFRbfARIgARIgAS8IUGF5ISZWkgRIgARIgAqL7wAJkAAJkIAXBKiwvBATK0kCJEACJECFxXeABEiABEjACwJUWF6IiZUkARIgARKgwuI7QAIkQAIk4AUBKiwvxMRKkgAJkAAJUGHxHSABEiABEvCCABWWF2JiJUmABEiABKiw+A6QAAmQAAl4QYAKywsxsZIkQAIkQAJUWHwHSIAESIAEvCBAheWFmFhJEiABEiABKiy+AyRAAiRAAl4QoMLyQkysJAmQAAmQABUW3wESIAESIAEvCFBheSEmVpIESIAESIAKi+8ACZAACZCAFwSosLwQEytJAiRAAiRAhcV3gARIgARIwAsCVFheiImVJAESIAESoMLiO0ACJEACJOAFASosL8TESpIACZAACVBh8R0gARIgARLwggAVlhdiYiVJgARIgASosPgOkAAJkAAJeEGACssLMbGSJEACJEACVFh8B0iABEiABLwgQIXlhZhYSRIgARIgASosvgMkQAIkQAJeEKDC8kJMrCQJkAAJkAAVFt8BEiABEiABLwhQYXkhJlaSBEiABEiACovvAAmQAAmQgBcEqLC8EBMrSQIkQAIkQIXFd4AESIAESMALAlRYXoiJlSQBEiABEqDC4jtAAiRAAiTgBQEqLC/ExEqSAAmQAAlQYfEdIAESIAES8IIAFZYXYmIlSYAESIAEqLD4DpAACZAACXhBgArLCzGxkiRAAiRAAlRYfAdIoAoJfPHFF/Lxxx/LvHnz5O+//5aZM2dK9+7dZbPNNstb2yeffFKmT58u6623nk23xhpryAYbbJA3D3+sDIFPPvlE3n77bVlmmWWkVatW0qZNm8rcOEN3ocLKkDD5KP4TePrpp2XMmDHy2WefyS+//CIzZsywx6ZNm8pKK60km2++uZx00kmy8cYbRx526tSpcuWVV9oG8YcffpBZs2bJ8ssvL0suuaTssMMOstdee8mee+4ZycOLyhF46KGH5JxzzpF//vlH5s6dK2uttZace+650qNHj4Iq8eGHH8r48eNl2rRpss4661jZbrPNNgWV4XViw0ACJNDoBGbPnm0uueQS06FDB6MNSt6/TTfd1IwePTqo8/3332/at2+fN89GG21kRo4cGeThSeUIvPPOO6Zjx44x+WjHo6BKfP3112a//fYzK6ywgi1LOyRmt912M99++21B5ficWHyuPOtOAlkgoF9T5ogjjjA6TBQ0avpFZRumTp06ReKdMmvbtq25/fbbzYMPPmigwFz82muvbTp37mx0SNCsttpqQTx+32qrrcy4ceOygMyrZ7jllltMy5YtI7KAPAYOHFjQc1x99dVmueWWi5VTSzKlwirolWFiEigtgTlz5pjevXtHGiEolkmTJhkdOrI3++uvv8zw4cNjimv99dc3G264YZAXymrUqFE2z+eff250CMro0GHwOxrJs88+u7QPwNIWSABfRa5D4Y5LLLFEIKsFFvB/gu222y5Wzuqrr27efffdtEV4n44Ky3sR8gF8JnD55ZebhRde2DZEGOpBr/vff/9NfKQBAwbEGizXALZr185g6AkBShCNm85fxdLvvPPOiWUzsjwEvvvuu0inwslLF8OYl19+OfVN0QHROauYPDHUq4tyUpfje0IqLN8lyPp7S+CBBx6wigWN2FJLLWWgvPKFiy++ONZgIW+LFi3MiBEjgqw33HCDwZCiaxzDR8ylMFSOwJQpU8yKK64YkwU6GD/++GPqijzxxBNWzmFZ4vywww5LXUYWElJhZUGKfAbvCGDeCkN/rgHq1avXAp+hX79+QXqXD0fMc82fPz/I37Nnz8R0SKsrBYN0PCk/gUGDBiXKolA56IrCxHLuuOOO8j9EFd2BCquKhMGq1A6Bm266KVjtte666wbDefkI7LvvvomN1oUXXhjJts8++ySmW3nllSNfYpFMvCgLga5duybK4vrrry/ofknltG7d2rzxxhsFleN7Yios3yXI+ntHAENBXbp0CRqyvn37pnqGLbfcMsjjvrDQaOlm1Ej+wYMHx4YEdQ+X2Xvvvc1vv/0WScuL8hHAopk111wzJrNVV13VvPrqq6lv/Omnn8YWz0D+WHBTa/Kkwkr92jAhCZSGwMSJEw1W9KHRwUKLNMuSMYSolhFijR9WAeYGtahg9t9/f7u0vVmzZmbbbbc1J554Yk2tJstl0hjXzz//vJ2bdJ0Ld8RCCd0UnrpKEyZMMOhwuPzumGYYOfVNPElISxcqfQYSqCQBXYYsOlkuiyyyiCy++OIyduxYWXrppfNWQSfdRb+QRHvUkXS77LKLaIMWicMFLCo8++yz1qoCzDnBskK5g65uFN1zJLrIoNy3Krj8uro62X333QvO15AMukhGdBtBrAhVNHLXXXfF4uuLOOuss0Q3lcd+1oU20r9//1h8piM8UaysJglkigCWImN/VdpwwgknxHrY2jCZoUOHpi2i7Omwkk3tFibWE3VtzD/ME6J+lQzamUh85ttuu62gaiTtv8KwYiHL4gu6YRUn5hdWprsjfLisENh6663lhRdeiDyOLoUXGMlt3rx5JL6xLnQjs6j5IIEtw2oLMDj7zDPPiFoFqUjVYC9QFY3gazocVlllFWsLMG09YAAZX4bvvfdeuBjRYUV57rnnrCHdyA8Zv6DCyriA+Xj+E4AhXBiwhbXvcNAJfWvRPRzXmOdqKkpuvPFGa12+MeuRdG9dISlDhgypyNAo7q/7r0QtXMg333wTqY5+6Vlr+osttlgkvr6LRx99VA4//HBRe4GRJBge1n18kbhauKDCqgUp8xm9JqD2AuXggw+W33//PfIcarVCHnvssUhcY19AuS5oPq4x6rjQQgtVtF7XXHON6DBu7FF1y4GoseJYfH0RsMx/1VVXxX6GZX5dSBOLz3oEFVbWJczn856AmmQStV4Re47rrrtO8BtD9RGAK5eHH344VjHI8ZhjjonF1xehWxlE56oiP+Nr8Z577pGacivyPwEqrMirwAsSqD4C8IH1+uuvRyqGOZlqnCuKVLJGLzB/hSFcOGsMB/gne+qpp0T3T4Wj6z3/6KOPRN2HCHxghQPmr+CoU7dEhKNr4pwKqybEzIf0lQAaLQz9YfI9HOC874MPPghH8bxKCOj+Kzt/lTvvBC/Qb731lqix41Q1ve++++z2h19//TWSftddd7ULN8KR2FLQpEmTcFQ2z6t4BSOrRgKZJvDHH38Y7Skb3YdlsNk3Kdx5550Grii09Yn8Ycl0moAl1N26dTM6fGR0VVmaLEzTQAJXXHFFRFZOdpBDIQE+0lze8BGOPsMBG5RhlxJ+0bD9IcuBli6yLF0+W9USgPNFHeozm2yyiYFfK+y1OfPMM2P1TfKlhMZr2LBhsbRJEbocPmj0jj/++KQkjCsxAf0CCpiHFQ2sjRQS8G6E8+McllHQyQkH+D1z6SDvQvb3hcvx4ZwKywcpsY6ZIoAeca43YDQ4W2yxhXnzzTeDZ/3zzz8TfSmphYwgTb4TlAVliLLxlTZmzJh8yflbCQjofimDTcpOgYSPurIv9R1gHzLpHYFZJ/jGCgd8Pbv7wEBylgMVVpaly2erOgK6NN2Ev3pcQ4Mj/CapyZ6gztOmTTNw9BdOg3N4mU0T8EXl8qLxgzNBhvISUHNYpmXLlgF3xx/HtF/FqOFpp52WWMaOO+4YeYCvvvoqcOyIjgx8oWU5UGFlWbp8tqoj8NJLLxmY1Qk3ZO4cwz2PPPJIUOc+ffokpkMP/rjjjgvSJZ1gvsopO92DZC666KKkZIwrMQFwdvLMPeo2hFR3Q7r6TFydd955kTLgpkY3Idt7wjgy5jyzHKiwsixdPlvVERg9erRZdNFFExs1fHnBKjsC3N1j+Ce30cM13ErkC7oSLeh1Iz0sus+ZMydfFv5WIgI77bRToswgh6OOOmqBd5kxY4ZdPJEkd8TdfffdkTLC94MftKwHKqysS5jPV1UEdCl64heW2pgzZ5xxRlBXuAdxjVZub1uXtBtdMh2kDZ/AoR/8Xrm8UG6TJk0KJ+F5mQhAJs5tjOMfPmIVXz63Imor0qjdwEB222+/fXDuynn88ceD2uMr2s1zwf0IhiOzHqiwsi5hPl9VEVBjtXZxhWuAdG+OadeunRk+fLitJ5w7nn766QbzEUijG4RNruNGfKFde+215qeffgqebfbs2QbDQyjLlY25LnX3EaThSXkJTJ48OeY408nCHfEVpPYFIxWBg0YshcdKUZcOw75J/s/cSkPkwbvj0qfd5hC5sYcX3DisEmcggUoSuOyyy0RdpMusWbPsbeGrCoZsl112WWs0dd68eYLNothgCiOnJ598spx//vkCn1guNG3aVLRXLTDdAwvpOvlufWXBlh8CrCEcdNBBokvlXRYey0wAMlJvzwu8CzZ9w2q7Ota0Viwgv+nTpweWS2AJY9CgQaLDwqIdE+vTLFyoKjb77jhjyChv1KhRonuxwskyeU6FVQGxwmjprbfeWrTxTbgogOuGYgJcUKDRY6geAl9++aXo4gqrhHKtebtaQonpUKA1fKpfSvLiiy9aY6rqWt0lSTy2bt1atGdulRzs2TFUjgDMMcGFSW6AQsk1r5SbBtdIB2ebutjGmmTSeUfRYWJrN1CHEpOySPv27eWUU06xnZPEBBmLpMKqgEB1qKYgg5elrJJuPhRY+9YVY6UslmU1kABMLcEQKnwa6X4r22PGUVcQ2o7NoYceahsu2Ax0AXlgBRxGVXVzqOjwofUsDOUGO3Xz588XmO054IAD7Beby8dj+QnokKx07txZZs6cGblZq1atbMfklVdesZ0OyAwdWKeA0KGEqxF8Ve2xxx4xY8ZqDcV6LZ44caL90kK+n3/+2d5D9+3J0Ucfbd2PRG6a4QsqrAoIFw7Yxo8fX4E7xW+h5mAiQ0nxFIxpTALoeet8hB0OatasmeikvaARg0Xu+gKcNr7//vuCRlI9F9v0UHQ6f1VzDv3qY1TpeF3wYL+KnCJy98fQ7GuvvSY6J2kVDmxD4qsaigwjJ7o3TnSlnx3CdXmSjvgqhwPPuro66xsLX90dOnSww8JJ6bMaR4VVZsnCAKau9rENTJlvFSsezuKOPPJIO2QQ+5ERJEACJSNwwQUXiO6RipXXv39/GTFiRCyeEcURoMIqjlvqXPAYesghhwQTqkkZdZ+M7TnpElXb24LLc/S8MJ6NHjjGtr///nvrCh09LBfgEwf/DOhl5wb04AYOHGjHwHN/4zUJkEBpCXTv3l0wbJcbdN+UHHjggbnRvC6SABVWkeDSZsPLCsWSFKBU4Ea7a9eu0rNnz6Qk9cbde++99usJY+K5AR5f+/XrJ/BKykACJFBeAvWNouD/Gx1WDAsylIYAFVZpOCaWgolx3SwoahAz8nuLFi2kV69eom4CippzgFv0Y4891s59RAr+/wLu1HX/jZ3MTfqdcSRAAqUjoPuvRI3O2lGRcKkYOZk6darovrlwNM8bQsDDvWPeVBl243TyPNjcp3Ky52ltiiU96JQpU+q1Bo3y1dmfwSZSBhIggcoQqM9+YO/evStTgRq6Cy1dlFHYffv2jSkr7GgvNsBdRK6ZHqcEcezYsaPRZbXFFs98JEACRRDQ+avY/zn+H0eOHFlEacySjwCHBPXNKlfo1KmTYP+FC5jPwiRsMQFLn7G/BkMMSQH7ONSwqmDfFQMJkEBlCGCJOuagc4f9sYF7woQJdvN3ZWpSG3ehwiqTnPEC9+jRQ5ypHKz0U8OVRW3gxYpBmNlJWoWE6mP/DTYnd+nSpUxPw2JJgASSCGD+Cgun5s6dG/kZHUf1ZyZNmjSJxPOigQTyfX7xt+IJYDhARWP/dEOoGTp0aFGFweusflkFZbky3bFt27ZGNyUXVTYzkQAJNIwAhvjd/2L4OGDAgIYVzNyJBPiFpW9ZOQLMrIwbN84WjS8gXYBRlC3BU0891a74wz6s3IBls0OGDOE+j1wwvCaBChFQ6+syduzYyN1g2PbSSy+1prUiP/Ci4QQS1RgjG0RArS8bfPmodOyfGqcsqjy4nAiX48rDUW3HmZtvvrmocpmJBEigNATUwHDwf+7+P9VkkoFfMobSE/gPAAD//8G5WZsAABBNSURBVO2dCWxUVRfHTzG2KCVRBAEFWRQQKlRSCBRCRJaiCSYIkSBokFIVECN7AFnEQAELhAAiUVlUXFAhLhXLviqIihsKCAErixUJKItQrMz3zsvHONO+luksb2ZufzcZ+97d7+8M8/fdd+694iGEnUBeXp6ncuXKHhHx1KxZ07Nz585yt/HWW295UlNT7Tq0Ht9PcnKyZ+bMmeWukwIQgEB4Cei/9Xr16tn/PqtVq+bp0KGDZ82aNeFthNq8BBL0yvoxJISRwJgxYyQnJ8eusXPnzrJ+/fpy1Z6bmyuTJk2Sb775pkS5hIQEGTZsmFiCJddee22JdCIgAAF3Caxbt042bdokiYmJ0q1bN0lPT3e3AxWoNQQrAsbWL/Brr70mSUlJ0rJlSxk6dGjArWzZskVGjBghu3fvdizTt29fmT9/vlj/N+eYTiQEIAABUwkgWDFkWRWprKwsxycr7ab+39uiRYukfv36MdRrugIBCEDAHQIIljucr9rKgQMHpH///rJjxw7HvO3atbPFqnnz5o7pREIAAhAwnQCCFQMWLigokIEDB8rq1asde9O2bVuZPXu2qGgRIAABCFRUAghWlC1//vx5GTJkiLz++uuOPUlLS5OpU6fKfffd55gezcjvv/9e1Akk1gNPpbFuIfoHgcAIIFiBcYpYrtGjR8usWbMc609JSZHJkyfLQw895Jgezch58+ZJdna2VKlSJZrduGrbRUVFtsdm7969r5qXDBCAQGwTQLCiaJ85c+bIyJEjHXugjhXPPPOM7cLumCGKkX///bftALJ9+/Yo9iLwppXz8OHDAy9ATghAICYJIFhRMotOAaqThVNQl/XMzEzvWi6nPNGOmzJliu26X6lSpbB1RZcEFp9ivNoywaulX758WV544QXhCStsZqIiCESNAIIVBfQffPCBLUinT58u0br+YPfr1y/sYlCioRAj/vnnH7FW+UudOnVCrOm/4oWFhWLtEGJHXBGucPxt3Ljxf41wBQEIxC0BBMtl023YsMH2CMzPz3dsWddavfPOO3LDDTc4phMJAQhAoKISQLBctPyuXbtkwIAB8tNPPzm22qpVK1m1apXUrVvXMb145KFDh2ThwoXyyy+/iLWHmf3Oq3ge7qND4Ny5c3LixInoNE6rEHAg0LBhQ4fY+IpCsFyy1969e+2pPqf9AbULjRo1Ep0qbNasWcA90rVbS5YssfM3bdq0VCEMuEIyho2AClbVqlXDVh8VQSAUAuog1b59+1CqiI2y1ktrQoQJHDlyxNO1a1e/Hdct63vvLY9Aj7XDRbl6cebMGY+1vshbh+VtWK7yZI4sgbNnz3pt42trrv/73sPCPRaWYEX2C+9S7eJSOxW2mT///NNjOVGU+uNVq1Ytz9q1a8vN57PPPvPUqFHDrvf222/3WO+9yl0HBSJHAMFy78cY4bs6a1MEiylB69seqaAu1bqWasGCBY5N6JTRsmXLpGfPno7pZUU+/fTT3nr13dfcuXPNeOQva9BxlKZTgo8++mgc9Ziumkxg1KhRRvw+IFgR/JbqLhW6BujixYuOrSxdulQee+wxx7SyItX9W0Vqz549drYnn3zS3hi3rDKkQQACEIh3AghWhCyoT1W6B+Dvv//u2IIewKgHPQYTrBNNRUVKXeN1Y1y9Dkb4gmmbMhCAAASiRQDBigD5FStWyIQJE+TgwYOOteuBjnoIY7ChR48e8uGHH9rF27RpYwtjly5dgq0uqHLa/nvvvRdUWTcL6ca3GRkZ9kGabrZLWxCAQAQIRO61c8Ws+dNPP/WkpqaW6mRhva/yqIdfsME6itujThbWV8H+qEOH28FaX+Tp1auXtw9X+hKLf5OTkz3W1KzbiGgPAhCIAAG8BMMIVT1xrDOrSv0ht9ZBeI4ePRp0i5cuXfJ06tTJW3+LFi08L730UtD1BVtQBbcsz8dYE66JEycGO1TKQQACMUSAKUHr1zUcQc+GUo/AzZs3O1anC4JXrlwpd955p2N6IJHqxLF48WI5duyYnT09PV3Gjx8v3bt3D6R4WPPoycgvvviiJCUlhbXeSFSWlZUlyooAAQjENwEEKwz2O3z4sOh7qdJODNbNV5cvXy6tW7cOujU9cTgnJ8fPiePBBx+0t3IKulIKQgACEIgjAghWiMY6efKkfaZVaScGN2jQQF555RXp3LlzUC1ZC1Dt+j/66CM/sbr55pvtpyt9qiNAAAIQqAgEEKwQrGy9U7K9AfXJxylUr17dFiv16itv0LVWumu7ur/rPoTFg7qz62JAy/mheBL3EIAABIwkgGCFYNbp06fbTzmlVaFPXYHudqC7Yhw/flz27dtnTy1u27bN3oVdn+Ccwr333ivTpk3j3YwTHOIgAAEjCSBYQZpV11npWirLY86xBl3/ozuo+wbL2Ub0qezChQv2R5+iNE6DCpMeilhQUCBFRUW+xRyvx40bJ9nZ2Y5pREIAAhAwkQCCFYRVrc1qZdCgQaLOFtEIevbV4MGD5eGHH45G87QJAQhAICoEEKwgsI8dO9Z+txRE0bAU0d0tZsyYIR07dgxLfVQCAQhAIB4IIFhBWMlaNCu67kqn8PTdk4YrfwOpTqcFExMTA8laIs+///4rmZmZYi2GLZFGBAQgAAGTCSBYQVh369at9rsofdd05R1U7dq1A65p//790qRJk4DzF8949913F4/iHgIQgIDxBBAs403MACEAAQiYQQDBMsOOjAICEICA8QQQLONNzAAhAAEImEEAwTLDjowCAhCAgPEEECzjTcwAIQABCJhBAMEyw46MAgIQgIDxBBAs403MACEAAQiYQQDBMsOOjAICEICA8QQQLONNzAAhAAEImEEAwTLDjowCAhCAgPEEECzjTcwAIQABCJhBAMEyw46MAgIQgIDxBBAs403MACEAAQiYQQDBMsOOjAICEICA8QQQLONNzAAhEF8E9MiegoIC2bhxo9xxxx1ldv7HH3+UlJSUMvNcvHhR7rnnnjLzkBgfBBCs+LATvYSA8QR++OEHWbRokRw7dkx+++03+5OUlCQJCQmljl3FqHLlyqWma0JhYaH06dPHPqW7rLrKrITEmCCAYMWEGegEBCougS+//FIWLFggO3fulCNHjsiFCxfCDqNevXqiB6/edtttYa+bCt0jgGC5x5qWIAABHwJnz56VOXPmyMcffyxff/21T0r4L3Xa8I033pCWLVuGv3JqdI0AguUaahqCAASuENi/f7+MHDlStm3bJmfOnLkSHbG/mZmZkpOTI9WqVYtYG1QceQIIVuQZ0wIEIOBD4Ntvv5Vhw4bJli1bfGL9L6+55hq55ZZb5NSpU3L+/Hn/RJ+7Vq1a2Xk0v4ZLly5Jfn6+fV23bl2pWbOmXc/48eOlTZs2djz/iV8CCFb82o6eQyDuCKhX3+DBg+0nq+Kd1/dLDRo0kPT0dOnQoYPceOONcvnyZdm8ebNs2rRJNmzYULyIJCcny/33328LoCaqd2GlSpVsoVKhq169ujRt2lTUeYMQ/wQQrPi3ISOAQFwQUDEZMGCA5OXllehvw4YNZfbs2dKjR48SaVciRo8eLbNmzbpy6/2blpZmP61VqVLFG8eFmQQQLDPtyqggEHMERo0aJW+//bYcP37cr286Vbdy5Uq59dZb/eKdbh544AHJzc31S9Kpw/fff99+MvNL4MY4AgiWcSZlQBCIPQIqKNOnT5fdu3f7dU6nAT///POAxEoLvvrqq/L444/71aE3S5YssZ/eSiQQYRQBBMsoczIYCMQeAV1b1atXL9H1Vr5BHSUWLlwoTzzxhG90mddfffWVZGRkyOnTp/3yTZw4UZ5//nm/OG7MI4BgmWdTRgSBmCKgYjJ16tQSferatausXbu2RHxZEevXr5eBAwfKr7/+6pdNHTlU/AhmE0CwzLYvo4NAVAnoNkvqSKFPRr6hRo0asmbNmnIv5F23bp307dtXTp486VudPPXUU/ZuGX6R3BhHAMEyzqQMCAKxQ+Ddd9+1FwgfPXrUr1Nt27aVHTt2+MUFcvPmm29KVlaW6B6CvmHKlCkyadIk3yiuDSSAYBloVIYEgVghoE9D6hlYPMycOVPGjBlTPPqq9xMmTJBp06b55dPNb3XT3P79+/vFc2MeAQTLPJsyIgjEBIE//vjDXtRbfJ/AOnXqiL6LatKkSbn72alTJ3sRsW/BWrVq2fsEdunSxTeaawMJIFgGGpUhQSAWCKhX4KBBg0q4suu6K92ZvbxB31t17NhRdLcM36CLjnUXjPr16/tGc20gAQTLQKMyJAjEAoH58+fL3Llz5dChQ37d0XVUL7/8sl9cIDe62/qQIUPk3Llzftl1a6bVq1f7xXFjJgEEy0y7MioIRJ2AOkgMHz5cdGrQN+iRIhpf3tC9e3f55JNP/IrpU5W+D+vdu7dfPDdmEkCwzLQro4JA1Ak8++yzkp2d7dcP3ZhWnTDKKzB79+6Vnj17yr59+/zqa926tS1i6iZPMJ8AgmW+jRkhBKJCYNWqVTJixAjvcR/aCd09Xd83tWjRolx9Gjt2rP0k5Vuodu3aoseGDB061Deaa4MJIFgGG5ehQSCaBLZv326vwdq1a5e3G3o+lbqgl7Uruzfz/y/0sMd+/fqVOJVYn7iWL18u1113XfEi3BtKAMEy1LAMCwLRJqC7svfp08fv7KvExERRIdOpvEDDI488IitWrJCioiJvEXW00MXC5anHW5iLuCWAYMWt6eg4BGKfgNPCYT2qXo8aCSQsXbrUdtD466+/vNnbtWsnejZWeZ7SvIW5iGsCCFZcm4/OQyC2CWzcuNF+j/Xdd995O9qsWTPZs2ePJCQkeOOcLpYtWybPPfec3zuwRo0ayeTJk+0pQqcyxJlNAMEy276MDgJRJ+B0UnC3bt1EnTKuv/76Ev07cOCAzJgxQ9QtvrCw0Juujhrjxo2zpxm9kVxUKAIIVoUyN4OFgPsE1CVdvQXz8vL8Gtf3T+3btxfdCFf3A9Qd3Q8fPiwHDx6UL774wi9vWlqavYegCh2h4hJAsCqu7Rk5BFwjoFOCOr23detWOXXqlF+7N910k+j6rOILjDVT8+bN7Y9OAzZu3NivHDcVjwCCVfFszoghEBUC+fn5kpubK4sXL5YTJ06InpVVWkhNTZWUlBT7sEbd8JYAASWAYPE9gAAEXCWgu1XolJ/u4v7zzz9LQUGBJCcn25+qVatKRkaGpKeny1133eVqv2gs9gkgWLFvI3oIAaMJqMu6vsPSkJSUZPRYGVxoBBCs0PhRGgIQgAAEXCKAYLkEmmYgAAEIQCA0AghWaPwoDQEIQAACLhFAsFwCTTMQgAAEIBAaAQQrNH6UhgAEIAABlwggWC6BphkIQAACEAiNAIIVGj9KQwACEICASwQQLJdA0wwEIAABCIRGAMEKjR+lIQABCEDAJQIIlkugaQYCEIAABEIjgGCFxo/SEIAABCDgEgEEyyXQNAMBCEAAAqERQLBC40dpCEAAAhBwiQCC5RJomoEABCAAgdAIIFih8aM0BCAAAQi4RADBcgk0zUAAAhCAQGgEEKzQ+FEaAhCAAARcIvA/5/DZrF6ptskAAAAASUVORK5CYII="
    }
   },
   "cell_type": "markdown",
   "id": "69a66c76",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "##### 4.6: Computing the z score for variant count for each population\n",
    "The z score will tell you how far a data point is from the mean. <br>\n",
    "Z score was computed using the formula: \n",
    "\n",
    "Z = (X - Œº) / ùùà\n",
    "\n",
    "Where:<br>\n",
    "Z - standard score <br>\n",
    "X - Observed value <br>\n",
    "Œº - Mean variant site count of all genomes (across populations included)<br>\n",
    "ùùà - Standard deviation computed across all genomes (all populations included)<br>\n",
    "\n",
    "Refer: [Extended Figure 4](https://www.nature.com/articles/nature15393/figures/8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "688e8824",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Compute number of variant sites per genome \n",
    "#Runtime - 1m 23s\n",
    "#Data processed - 4.78TB\n",
    "\n",
    "z_score_sql<-\"WITH cte1 AS (SELECT Population,Sample,\n",
    "COUNT(reference_name) as variant_site,\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v,UNNEST(v.alternate_bases) AS Alternate, \n",
    "UNNEST(v.call) AS call\n",
    "INNER JOIN `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`s\n",
    "ON call.name=s.Sample\n",
    "WHERE EXISTS(SELECT gt FROM UNNEST(call.genotype) gt WHERE gt>0 ) AND reference_name NOT IN ('X','Y','MT')\n",
    "GROUP BY 1,2), cte2 AS (SELECT AVG(variant_site) AS total_mean, STDDEV(variant_site) AS total_sd FROM cte1), \n",
    "cte3 AS (SELECT Population, AVG(variant_site) AS sub_population_mean FROM cte1 GROUP BY 1) \n",
    "SELECT Population, (sub_population_mean-total_mean)/total_sd AS Z_score FROM cte2 join cte3 on 1=1 \n",
    "GROUP BY 1,2 ORDER BY Z_score DESC\"\n",
    "\n",
    "z_score<-dbGetQuery(gcp_account, z_score_sql)\n",
    "z_score"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31db2935",
   "metadata": {},
   "source": [
    "For the computation of the z score, the query has been broken into 4 sections:<br>\n",
    "- Subquery 1: Join the variant and sample table based on sample id <br>\n",
    "- Subquery 2: Compute the population mean and standard deviation across all samples (and sub-population groups)<br>\n",
    "- Subquery 3: Compute the sample mean across variant sites per sub-population group <br>\n",
    "- Finally, calulate the z score using the formula <br>\n",
    "\n",
    "The output dataframe has 2 colums, the sub-population and the z-score calculated and will be used as the input dataframe to generate the below graph.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a7db42a",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Runtime - 0.7s\n",
    "\n",
    "#Make plots wider on jupyter notebooks\n",
    "options(repr.plot.width=17, repr.plot.height=8)\n",
    "\n",
    "#Set the order of the labels \n",
    "z_score$Population<-factor(z_score$Population, levels = unique(z_score$Population))\n",
    "\n",
    "#Generating the graph \n",
    "ggplot(z_score, aes(y=Z_score,x=Population))+\n",
    "geom_bar(aes(fill=Population), stat = \"identity\", position = \"dodge\")+\n",
    "theme(axis.text.x = element_blank(),axis.ticks.x = element_blank())+\n",
    "scale_y_continuous(breaks = c(-2.0,-1.5,-1.0,-0.5,0,0.5,1.0,1.5,2.0),minor_breaks = NULL)+\n",
    "labs(title = \"Z score for variant count for each population\",y = \"Standardized number of sites/genome\")+\n",
    "theme(plot.title = element_text(size=20),axis.text=element_text(size=14),axis.title=element_text(size=18))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bf51290",
   "metadata": {},
   "source": [
    "##### 4.7: Computation of allele frequency by changing syntax from Legacy SQL to Standard SQL \n",
    "Some important fucntions that have been used to compute allele frequency on the 1000 Genomes dataset\n",
    "- __```LENGTH()``` function:__ \n",
    "    * This function computes the length of the array element provided\n",
    "    * In this case we have used the ```alt``` as an input to the ```LENGTH()``` function\n",
    "    * The function returns the length of the string in bytes. The equivalent of this in legacy SQL is ```POSITION()``` \n",
    "    * Usage: ```LENGTH(alt) AS alt``` \n",
    "    \n",
    "    <br>\n",
    "    <br>\n",
    "    \n",
    "- __```CAST()``` function :__\n",
    "    * This function converts the value of any data type to the specified data type\n",
    "    * We have used this function to convert the result obtained from ```LENGTH()``` into integer type\n",
    "    * The equivalent of this in legacy SQL is ```INTEGER() WITHIN RECORD```  \n",
    "    * Usage: ```SUM(CAST(genotype = alt AS INT64)) AS alt_count``` - This converts the values of ```alt``` from bytes to integer type\n",
    "\n",
    "References: \n",
    "- [Million Veterean Program Codelabs](https://github.com/StanfordBioinformatics/mvp_aaa_codelabs/blob/master/sql)\n",
    "- [Google genomics - BigQuery examples](https://github.com/deflaux/bigquery-examples/tree/master/1000genomes/sql)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f325ed8e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Validating allele frequency computation code on 1000Genomes dataset\n",
    "#Runtime - ~20 min \n",
    "#Data processed - 3.06 TB - costs $10.30  \n",
    "\n",
    "allele_frequency_sql<-\"WITH cte1 AS (SELECT reference_name, start_position, reference_bases, AF AS AF_thousand_genome , LENGTH(alt) AS alt,\n",
    "         name,genotype, \n",
    "          FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v, \n",
    "          UNNEST(v.call) AS call, UNNEST(call.genotype) AS genotype, UNNEST(alternate_bases) AS alternate_bases\n",
    "          ),cte2 AS (SELECT AF_thousand_genome, reference_name,reference_bases, alt,start_position,\n",
    "                     SUM(CAST(genotype = 0 AS INT64)) AS ref_count, \n",
    "                     SUM( CAST(genotype = alt AS INT64)) AS alt_count,\n",
    "                     FROM cte1 GROUP BY 1,2,3,4,5) SELECT reference_name, start_position, reference_bases,\n",
    "                                          AF_thousand_genome,\n",
    "                                          SUM(alt_count)/NULLIF(SUM(ref_count)+SUM(alt_count),0) AS alt_freq,\n",
    "                                          SUM(ref_count)/NULLIF(SUM(ref_count)+SUM(alt_count),0) AS ref_freq,\n",
    "                                          SUM(ref_count)+SUM(alt_count) AS num_sample_alleles,\n",
    "                                          SUM(ref_count) AS ref_cnt,\n",
    "                                          SUM(alt_count) AS alt_cnt,\n",
    "                                          FROM cte2 GROUP BY 1,2,3,4\"\n",
    "\n",
    "allele_freqeuency<-dbGetQuery(gcp_account, allele_frequency_sql)\n",
    "head(allele_freqeuency)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Slideshow",
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
