{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b7764602",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# 1000 Genomes analysis on BigQuery"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8586947e-fada-4ded-a14d-752ac1493719",
   "metadata": {},
   "source": [
    "### 0. Verily Workbench Introduction\n",
    "\n",
    "This notebook, from the [VA Big Data Genomics Group](https://github.com/va-big-data-genomics), shows an example of using R for analysis of the 1000 Genomes BigQuery dataset. (It also illustrates how notebooks can be used to help reproduce publication results-- in this case for the 1000 Genomes phase 3 publication (https://doi.org/10.1038/nature15393)).\n",
    "\n",
    "**This notebook is intended to be run on an R-based Verily Workbench workspace cloud environment**.\n",
    "The notebook has been modified to automatically grab the Workbench workspace's underlying GCP project ID, but is otherwise unchanged.\n",
    "\n",
    "**The original notebook source is here:  \n",
    "https://github.com/va-big-data-genomics/variant-analysis-on-bigquery/blob/main/1000g_Analysis.ipynb**\n",
    "\n",
    "Note that estimates of data processed are included below, which can be used to estimate costs.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da31a4cc",
   "metadata": {},
   "source": [
    "### 1. Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb90f658",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "Authors: Prathima Vembu <br>\n",
    "Notebook kernel: R (https://irkernel.github.io/installation/)<br>\n",
    "Recommended Google Cloud Platform machine type: n1-standard-1 <br>\n",
    "Predicted runtime: ~30 minutes <br>\n",
    "Predicted cost for analysis <br>\n",
    "1. Variant optimized table : 83.40 USD per one month (~100 USD including allele frequency computation) <br>\n",
    "2. Sample optimized table : 0.00 USD per one month first \n",
    "*Note: 1 TB of query data processed per month is free* <br>\n",
    "\n",
    "This notebook includes methods for reproducing several figures in the 1000 Genomes phase 3 publication (https://doi.org/10.1038/nature15393)<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f056c5ee",
   "metadata": {},
   "source": [
    "##### 1.1: Comparing variant optimized table with sample optimized table\n",
    "||Variant optimized|Sample optimized|\n",
    "|---|---|---|\n",
    "|User specified |No, generated by default|Yes, use ```--sample_lookup_optimized_output_table``` parameter|\n",
    "|Number of tables created |25 (23 chromosome, sample_info, residual_info)|25 (23 chromosome, sample_info, residual_info)|\n",
    "|Analysis |Variant lookup query - Find all variants within a enomic region|Sample lookup query - Find all variants in Sample 'Z'|\n",
    "|Storage cost|No additional storage cost|Additional storage cost (for 25 tables), recovered when running queries for sample lookup analysis|\n",
    "|Call field|Call field *not* unnested |Call field unnested|\n",
    "|New field|No new column added|New column ```sample_id``` added which has the same value as ```call.sample_id``` column|\n",
    "|1KG table size|1000_genomes_phase_3_variants_20150220 - 5.38 TB|1000_genomes_phase_3_optimized_schema_variants_20150220 - 1.94 TB|"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da46a038",
   "metadata": {},
   "source": [
    "##### 1.2: Table location \n",
    "Variant optimized table : ```bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220``` <br>\n",
    "Sample optimized table : ```bigquery-public-data.human_genome_variants.1000_genomes_phase_3_optimized_schema_variants_20150220```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "065025bd",
   "metadata": {},
   "source": [
    "##### 1.3: Query cost comparison \n",
    "\n",
    "| Analysis | Variant table(1000Genomes variant table)| Data processed | Sample table(1000Genomes sample optimized table) | Data processed |\n",
    "|---|---|---|---|---|\n",
    "|Distinct sample_id’s|3.70 USD |1.74 TB | 0.46 USD|0.092 TB|\n",
    "|Number of mapped and missing ID’s|3.70 USD|1.74 TB|0.46 USD|0.092 TB|\n",
    "|Number of variant sites per sample| 19.00 USD|4.8 TB|1.27 USD|0.255 TB|\n",
    "|Average singletons|19.00 USD|4.8 TB|1.28 USD|0.256 TB|\n",
    "|Average singletons and rare variants |19.00 USD|4.8 TB|1.28 USD|0.256 TB|\n",
    "|Distribution of Allele Frequency|0.00 USD|0.001 TB|0.00 USD|0.001 TB|\n",
    "|Z-score computation|19.00 USD|4.8 TB|1.27 USD|0.255 TB|\n",
    "|Total|83.4 USD|22.68 TB|6.02 USD|1.207 TB|\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddd3be80",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### 2. Installing necessary packages in R\n",
    "Before starting analysis, install and load the below packages in R. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32710832",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    },
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#helps connecting R to DBMS\n",
    "install.packages(\"DBI\") \n",
    "#helps to work with data on cloud storage\n",
    "install.packages(\"bigrquery\") \n",
    "#package to perform multiple data manipulation functions\n",
    "install.packages(\"pillar\") \n",
    "#package to restructure and aggregate dataframes\n",
    "install.packages(\"reshape2\") \n",
    "#package for creating graphics \n",
    "install.packages(\"ggplot2\") \n",
    "\n",
    "library(DBI)\n",
    "library(bigrquery)\n",
    "library(pillar)\n",
    "library(reshape2)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cf04afc",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### 3. Authenticate and connect to BigQuery \n",
    "Connect to the Google BigQuerydatabase account using ```dbconnect```. Replace ```BILLING_ACCOUNT``` with the name of the Google Cloud project you want charges to be billed to."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8da9566f-dd2a-49ca-8973-98ce36572cb4",
   "metadata": {},
   "source": [
    "If you're running this notebook on a Verily Workbench workspace cloud environment, use the following cell to automatically grab the project ID. Otherwise, you will first need to set the `PROJECT_ID` manually."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "018f9d35-c96d-445e-bbbc-a0b734b4ad9b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "PROJECT_ID <- Sys.getenv('GOOGLE_CLOUD_PROJECT')\n",
    "PROJECT_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd87d095",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#dbconnect creates a connection to the DBMS\n",
    "gcp_account<-dbConnect(bigquery(),\"bigquery-public-data\",\n",
    "                       dataset=\"human_genome_variants\", \n",
    "                       billing = PROJECT_ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc21eb2f",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### 4. Analysis\n",
    "\n",
    "##### 4.1: Check to confirm the number of samples present in the BigQuery table (from the paper, n=2504). <br>\n",
    "Two tables are available for the 1000 Genomes project which can be found under *bigquery-public-data.human_genome_variants* <br>\n",
    "1. 1000_genomes_phase_3_variants_20150220 : The Variant table<br>\n",
    "2. 1000_genomes_sample_info : The Sample table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27b9d41f",
   "metadata": {
    "scrolled": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Number of samples in the Variant table\n",
    "#Runtime - 20s\n",
    "#Data processed - 0.092 TB\n",
    "variant_sample_count_sql<-\"SELECT\n",
    "COUNT(distinct call.name) AS Sample_count\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v, UNNEST(v.call) AS call\"\n",
    "\n",
    "variant_sample_count<-dbGetQuery(gcp_account,variant_sample_count_sql)\n",
    "variant_sample_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c9bb8a7",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Number of samples in the Sample table \n",
    "#Runtime - 1.8s\n",
    "#Data processed - 30.76KB\n",
    "sample_table_count_sql<-\"SELECT\n",
    "COUNT(Sample) AS Sample_count\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`\"\n",
    "\n",
    "sample_table_count<-dbGetQuery(gcp_account, sample_table_count_sql)\n",
    "sample_table_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f9b2039",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "Given that both these tables have different number of distinct sample ID's, it is important to make sure that the list of sample ID's in the ***variant table*** is a subset of the ID's from the ***sample table*** to prevent data loss. <br>\n",
    "The query below computes the number of mapped ID's and number of missing ID's. All ID's from the ***variant table*** match with entries in the ***sample table***. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a77e377",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Compute the number of mapped ID's and the number of missing ID's\n",
    "#Runtime - 20s\n",
    "#Data processed - 1.74TB\n",
    "mapped_missing_values_sql<-\"WITH cte1 AS (\n",
    "SELECT DISTINCT call.name AS call_name\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v,UNNEST(v.call) AS call\n",
    "), cte2 AS (SELECT distinct Sample AS sample_name \n",
    " FROM `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`\n",
    "),cte3 AS(SELECT call_name, sample_name \n",
    "  FROM cte1 \n",
    "  RIGHT JOIN cte2 \n",
    "  ON cte1.call_name=cte2.sample_name\n",
    ")\n",
    "SELECT SUM(CASE WHEN call_name IS NULL THEN 1 ELSE 0 END) AS missing_values, \n",
    "SUM(CASE WHEN call_name IS NOT NULL THEN 1 ELSE 0 END) AS mapped_values FROM cte3\"\n",
    "\n",
    "mapped_missing_values<-dbGetQuery(gcp_account, mapped_missing_values_sql)\n",
    "mapped_missing_values\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25aa1138",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "##### 4.2: Number of variant sites per genome \n",
    "Refer: [Figure 1b](https://www.nature.com/articles/nature15393/figures/1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cc05f4d",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Compute the number of variant sites per genome\n",
    "#Runtime - 1.9s\n",
    "#Data processed - 4.8 TB\n",
    "\n",
    "variant_sites_per_genome_sql<-\"SELECT s.Population,s.Sample,COUNT(reference_name) as variant_count,\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v, UNNEST(v.call) AS call\n",
    "INNER JOIN `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`s\n",
    "ON call.name=s.Sample\n",
    "WHERE EXISTS (SELECT gt FROM UNNEST(call.genotype) gt WHERE gt > 0 AND reference_name NOT IN ('X','Y','MT'))\n",
    "GROUP BY 1,2 ORDER BY variant_count DESC\"\n",
    "\n",
    "variant_sites_per_genome<-dbGetQuery(gcp_account, variant_sites_per_genome_sql)\n",
    "head(variant_sites_per_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb75711f",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Runtime - 2.5s\n",
    "\n",
    "#Make plots wider on jupyter notebooks\n",
    "options(repr.plot.width=15, repr.plot.height=8)\n",
    "\n",
    "#Set the order of the labels \n",
    "variant_sites_per_genome$Population<-factor(variant_sites_per_genome$Population,levels = unique(variant_sites_per_genome$Population))\n",
    "\n",
    "#Generate the plot\n",
    "ggplot(variant_sites_per_genome, aes(y=variant_count,x=Population, color=Population))+\n",
    "geom_point()+theme(axis.text.x = element_blank(),axis.ticks.x = element_blank())+\n",
    "scale_y_continuous(breaks = c(3800000,4000000,4200000,4400000,4600000,4800000,5000000,5200000,5400000,5600000),\n",
    "                   minor_breaks = NULL, labels = function(l) {paste0(round(l/1e6,1),\"M\")})+\n",
    "labs(title = \"Variant sites per genome across populations\",x = \"Individual\",y = \"Variant sites\")+\n",
    "theme(plot.title = element_text(size=22),axis.text=element_text(size=14),axis.title=element_text(size=18))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77328be7",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "##### 4.3: The average number of singletons per genome\n",
    "Refer: [Figure 1c](https://www.nature.com/articles/nature15393/figures/1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fee7d7af",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Compute the number of singletons per genome\n",
    "#Runtime - 1.5s\n",
    "#Data processed - 4.8TB\n",
    "singletons_per_genome_sql<-\"WITH cte1 AS (SELECT s.Population,Sample,\n",
    "COUNT(reference_name) as variant_sites,COUNTIF(AC=1) AS Singletons,\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v,\n",
    "UNNEST(v.alternate_bases) AS Alternate, UNNEST(v.call) AS call\n",
    "INNER JOIN `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`s\n",
    "ON call.name=s.Sample\n",
    "WHERE EXISTS(SELECT gt FROM UNNEST(call.genotype) gt WHERE gt>0 ) AND reference_name NOT IN ('X','Y','MT')\n",
    "GROUP BY 1,2)\n",
    "SELECT Population, AVG(Singletons) AS Average_Singletons FROM cte1 GROUP BY 1 ORDER BY 2 DESC\"\n",
    "\n",
    "singletons_per_genome<-dbGetQuery(gcp_account, singletons_per_genome_sql)\n",
    "head(singletons_per_genome)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "531b37b4",
   "metadata": {},
   "source": [
    "The output table is sorted in a descending order by the ***Average_Singletons*** column. Output dataframe has 2 columns, the ***Population*** group and the ***Average_Singletons*** computed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aed2d7bd",
   "metadata": {
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Runtime - 0.9s\n",
    "\n",
    "#Make plots wider on jupyter notebooks\n",
    "options(repr.plot.width=15, repr.plot.height=8)\n",
    "\n",
    "#Setting the labels, AC=1\n",
    "singletons_per_genome$Population<-factor(singletons_per_genome$Population,levels = unique(singletons_per_genome$Population))\n",
    "\n",
    "#Generate the plot, AC=1\n",
    "ggplot(singletons_per_genome,aes(y=Average_Singletons,x=Population))+\n",
    "geom_bar(aes(fill=Population),stat = \"identity\",position=\"dodge\")+\n",
    "scale_y_continuous(breaks = c(0,2000,4000,6000,8000,10000,12000,14000,16000,18000,20000,22000),\n",
    "                   minor_breaks = NULL, labels = function(l) {paste0(round(l/1e3,1),\"K\")})+\n",
    "labs(title = \"Average number of singletons per genome\",x = \"Population\",y = \"Singletons per genome\")+\n",
    "theme(plot.title = element_text(size=22),axis.text=element_text(size=14),axis.title=element_text(size=18))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "358d47bd",
   "metadata": {},
   "source": [
    "##### 4.4: Distribution of allele frequency \n",
    "Refer: Extended Figure 3a\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d7a4a03",
   "metadata": {},
   "source": [
    "Allele frequency has been extracted from the 1KG table and split into 20 buckets. This data is then plotted as a bar graph to see th allele frequency distribution. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5bf815e",
   "metadata": {
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Bucketing the allele frequency\n",
    "#Runtime - 16s\n",
    "#Data processed - 650.58MB\n",
    "af_sql<-\"SELECT COUNT(AF) AF_Count, AF_bucket, \n",
    "FROM (SELECT  AF, CASE WHEN AF=0 THEN 0\n",
    "                       WHEN AF>0 AND AF<=0.05 THEN 1\n",
    "                       WHEN AF>0.05 AND AF<=0.1 THEN 2\n",
    "                       WHEN AF>0.1 AND AF<=0.15 THEN 3\n",
    "                       WHEN AF>0.15 AND AF<=0.2 THEN 4\n",
    "                       WHEN AF>0.2 AND AF<=0.25 THEN 5\n",
    "                       WHEN AF>0.25 AND AF<=0.3 THEN 6\n",
    "                       WHEN AF>0.3 AND AF<=0.35 THEN 7\n",
    "                       WHEN AF>0.35 AND AF<=0.4 THEN 8\n",
    "                       WHEN AF>0.4 AND AF<=0.45 THEN 9\n",
    "                       WHEN AF>0.45 AND AF<=0.5 THEN 10\n",
    "                       WHEN AF>0.5 AND AF<=0.55 THEN 11\n",
    "                       WHEN AF>0.55 AND AF<=0.6 THEN 12\n",
    "                       WHEN AF>0.6 AND AF<=0.65 THEN 13\n",
    "                       WHEN AF>0.65 AND AF<=0.7 THEN 14\n",
    "                       WHEN AF>0.7 AND AF<=0.75 THEN 15\n",
    "                       WHEN AF>0.75 AND AF<=0.8 THEN 16\n",
    "                       WHEN AF>0.8 AND AF<=0.85 THEN 17\n",
    "                       WHEN AF>0.85 AND AF<=0.9 THEN 18\n",
    "                       WHEN AF>0.9 AND AF<=0.95 THEN 19\n",
    "                       WHEN AF>0.95 AND AF<=1.0 THEN 20\n",
    "                       ELSE -1 END AS AF_bucket\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v,UNNEST(v.alternate_bases) AS Alternate )\n",
    "GROUP BY AF_bucket\n",
    "ORDER BY AF_bucket\"\n",
    "\n",
    "af<-dbGetQuery(gcp_account, af_sql)\n",
    "af"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "868c9583",
   "metadata": {
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Runtime - 0.4s\n",
    "\n",
    "##REMOVING FIRST ROW - AF=0\n",
    "#Make plots wider on jupyter notebooks\n",
    "options(repr.plot.width=23, repr.plot.height=8)\n",
    "\n",
    "#Removing AF=0 row \n",
    "af_mod = af[-1,]\n",
    "\n",
    "#Generating the graph for variant tables \n",
    "ggplot(af_mod, aes(x=AF_bucket, y=AF_Count))+\n",
    "geom_bar(stat = \"identity\",fill=\"red\")+\n",
    "labs(title = \"Variant table - Allele Frequency distribution\",x = \"Allele frequency range\",y = \"Count(allele frequency)\")+\n",
    "theme(plot.title = element_text(size=22),axis.text=element_text(size=14),axis.title=element_text(size=18), \n",
    "      legend.title = element_blank(),legend.text = element_blank())+\n",
    "scale_x_continuous(breaks=c(1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20), minor_breaks=NULL,labels=c(\"0.05\",\"0.1\",\"0.15\", \"0.2\", \"0.25\", \"0.3\", \"0.35\", \"0.4\", \"0.45\", \n",
    "      \"0.5\", \"0.55\", \"0.6\", \"0.65\", \"0.7\", \"0.75\", \"0.8\", \"0.85\", \"0.9\", \"0.95\",\"1.0\"))+\n",
    "geom_text(aes(label = scales::comma(AF_Count)),vjust=-0.2,size=5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d50310a",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "##### 4.5 : Number of rare variants and singletons per genome <br>\n",
    "Refer: [Extended Figure 3b](https://www.nature.com/articles/nature15393/figures/7)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5dbbb80",
   "metadata": {
    "scrolled": true,
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Compute the number of rare variants and singletons per genome\n",
    "#Runtime - 43s\n",
    "#Data processed - 4.8TB\n",
    "rare_var_and_singletons_per_genome_sql<-\"WITH cte1 AS (SELECT Population,Sample,\n",
    "COUNT(reference_name) as variant_sites,\n",
    "COUNTIF(AF<0.005) AS Rare_Variants,\n",
    "COUNTIF(AC=1) AS Singletons,\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v,\n",
    "UNNEST(v.alternate_bases) AS Alternate, UNNEST(v.call) AS call\n",
    "INNER JOIN `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`s\n",
    "ON call.name=s.Sample\n",
    "WHERE EXISTS(SELECT gt FROM UNNEST(call.genotype) gt WHERE gt>0 ) AND reference_name NOT IN ('X','Y','MT')\n",
    "GROUP BY 1,2) SELECT Population, AVG(Rare_variants) AS Average_rare_variants, AVG(Singletons) AS Average_singletons \n",
    "FROM cte1 GROUP BY 1 ORDER BY Average_rare_variants,Average_singletons DESC\"\n",
    "\n",
    "rare_var_and_singletons_per_genome<-dbGetQuery(gcp_account, rare_var_and_singletons_per_genome_sql)\n",
    "head(rare_var_and_singletons_per_genome)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "561aa8f7",
   "metadata": {},
   "source": [
    "The output table contains 3 colums, the ***Population***, ***Average_rare_variants*** computed and the ***Average_singletons*** computed. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c208842",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Runtime - 1.5s\n",
    "\n",
    "#Make plots wider on jupyter notebooks\n",
    "options(repr.plot.width=15, repr.plot.height=8)\n",
    "\n",
    "#Set the order of the labels \n",
    "rare_var_and_singletons_per_genome$Population<-factor(rare_var_and_singletons_per_genome$Population,levels = unique(rare_var_and_singletons_per_genome$Population))\n",
    "\n",
    "#Generating the graph \n",
    "ggplot(rare_var_and_singletons_per_genome,aes(x=Population, y=Average_rare_variants))+\n",
    "geom_bar(stat = \"identity\", position = \"stack\", fill=\"grey\")+\n",
    "geom_bar(data = subset(rare_var_and_singletons_per_genome,variable=Average_singletons),\n",
    "         aes(x=Population, y=Average_singletons, fill=Population),stat = \"identity\")+\n",
    "scale_y_continuous(breaks = c(0,20000,40000,60000,80000,100000,120000,140000,160000,180000,200000,220000,\n",
    "                              240000,260000,280000),labels = function(l) {paste0(round(l/1e3,1),\"K\")})+\n",
    "labs(title = \"Average number of rare variants (grey) with average number of singletons (coloured)\",\n",
    "     y = \"Rare variants per genome\")+\n",
    "theme(plot.title = element_text(size=20),axis.text=element_text(size=14),axis.title=element_text(size=18))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69a66c76",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "##### 4.6: Computing the z score for variant count for each population\n",
    "The z score will tell you how far a data point is from the mean. <br>\n",
    "Z score was computed using the formula: \n",
    "\n",
    "Z = (X - μ) / 𝝈\n",
    "\n",
    "Where:<br>\n",
    "Z - standard score <br>\n",
    "X - Observed value <br>\n",
    "μ - Mean variant site count of all genomes (across populations included)<br>\n",
    "𝝈 - Standard deviation computed across all genomes (all populations included)<br>\n",
    "\n",
    "Refer: [Extended Figure 4](https://www.nature.com/articles/nature15393/figures/8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "688e8824",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Compute number of variant sites per genome \n",
    "#Runtime - 1m 23s\n",
    "#Data processed - 4.78TB\n",
    "\n",
    "z_score_sql<-\"WITH cte1 AS (SELECT Population,Sample,\n",
    "COUNT(reference_name) as variant_site,\n",
    "FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v,UNNEST(v.alternate_bases) AS Alternate, \n",
    "UNNEST(v.call) AS call\n",
    "INNER JOIN `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`s\n",
    "ON call.name=s.Sample\n",
    "WHERE EXISTS(SELECT gt FROM UNNEST(call.genotype) gt WHERE gt>0 ) AND reference_name NOT IN ('X','Y','MT')\n",
    "GROUP BY 1,2), cte2 AS (SELECT AVG(variant_site) AS total_mean, STDDEV(variant_site) AS total_sd FROM cte1), \n",
    "cte3 AS (SELECT Population, AVG(variant_site) AS sub_population_mean FROM cte1 GROUP BY 1) \n",
    "SELECT Population, (sub_population_mean-total_mean)/total_sd AS Z_score FROM cte2 join cte3 on 1=1 \n",
    "GROUP BY 1,2 ORDER BY Z_score DESC\"\n",
    "\n",
    "z_score<-dbGetQuery(gcp_account, z_score_sql)\n",
    "z_score"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31db2935",
   "metadata": {},
   "source": [
    "For the computation of the z score, the query has been broken into 4 sections:<br>\n",
    "- Subquery 1: Join the variant and sample table based on sample id <br>\n",
    "- Subquery 2: Compute the population mean and standard deviation across all samples (and sub-population groups)<br>\n",
    "- Subquery 3: Compute the sample mean across variant sites per sub-population group <br>\n",
    "- Finally, calulate the z score using the formula <br>\n",
    "\n",
    "The output dataframe has 2 colums, the sub-population and the z-score calculated and will be used as the input dataframe to generate the below graph.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a7db42a",
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    },
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Runtime - 0.7s\n",
    "\n",
    "#Make plots wider on jupyter notebooks\n",
    "options(repr.plot.width=17, repr.plot.height=8)\n",
    "\n",
    "#Set the order of the labels \n",
    "z_score$Population<-factor(z_score$Population, levels = unique(z_score$Population))\n",
    "\n",
    "#Generating the graph \n",
    "ggplot(z_score, aes(y=Z_score,x=Population))+\n",
    "geom_bar(aes(fill=Population), stat = \"identity\", position = \"dodge\")+\n",
    "theme(axis.text.x = element_blank(),axis.ticks.x = element_blank())+\n",
    "scale_y_continuous(breaks = c(-2.0,-1.5,-1.0,-0.5,0,0.5,1.0,1.5,2.0),minor_breaks = NULL)+\n",
    "labs(title = \"Z score for variant count for each population\",y = \"Standardized number of sites/genome\")+\n",
    "theme(plot.title = element_text(size=20),axis.text=element_text(size=14),axis.title=element_text(size=18))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bf51290",
   "metadata": {},
   "source": [
    "##### 4.7: Computation of allele frequency by changing syntax from Legacy SQL to Standard SQL \n",
    "Some important fucntions that have been used to compute allele frequency on the 1000 Genomes dataset\n",
    "- __```LENGTH()``` function:__ \n",
    "    * This function computes the length of the array element provided\n",
    "    * In this case we have used the ```alt``` as an input to the ```LENGTH()``` function\n",
    "    * The function returns the length of the string in bytes. The equivalent of this in legacy SQL is ```POSITION()``` \n",
    "    * Usage: ```LENGTH(alt) AS alt``` \n",
    "    \n",
    "    <br>\n",
    "    <br>\n",
    "    \n",
    "- __```CAST()``` function :__\n",
    "    * This function converts the value of any data type to the specified data type\n",
    "    * We have used this function to convert the result obtained from ```LENGTH()``` into integer type\n",
    "    * The equivalent of this in legacy SQL is ```INTEGER() WITHIN RECORD```  \n",
    "    * Usage: ```SUM(CAST(genotype = alt AS INT64)) AS alt_count``` - This converts the values of ```alt``` from bytes to integer type\n",
    "\n",
    "References: \n",
    "- [Million Veterean Program Codelabs](https://github.com/StanfordBioinformatics/mvp_aaa_codelabs/blob/master/sql)\n",
    "- [Google genomics - BigQuery examples](https://github.com/deflaux/bigquery-examples/tree/master/1000genomes/sql)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f325ed8e",
   "metadata": {
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#Validating allele frequency computation code on 1000Genomes dataset\n",
    "#Runtime - ~20 min \n",
    "#Data processed - 3.06 TB - costs $10.30  \n",
    "\n",
    "allele_frequency_sql<-\"WITH cte1 AS (SELECT reference_name, start_position, reference_bases, AF AS AF_thousand_genome , LENGTH(alt) AS alt,\n",
    "         name,genotype, \n",
    "          FROM `bigquery-public-data.human_genome_variants.1000_genomes_phase_3_variants_20150220`v, \n",
    "          UNNEST(v.call) AS call, UNNEST(call.genotype) AS genotype, UNNEST(alternate_bases) AS alternate_bases\n",
    "          ),cte2 AS (SELECT AF_thousand_genome, reference_name,reference_bases, alt,start_position,\n",
    "                     SUM(CAST(genotype = 0 AS INT64)) AS ref_count, \n",
    "                     SUM( CAST(genotype = alt AS INT64)) AS alt_count,\n",
    "                     FROM cte1 GROUP BY 1,2,3,4,5) SELECT reference_name, start_position, reference_bases,\n",
    "                                          AF_thousand_genome,\n",
    "                                          SUM(alt_count)/NULLIF(SUM(ref_count)+SUM(alt_count),0) AS alt_freq,\n",
    "                                          SUM(ref_count)/NULLIF(SUM(ref_count)+SUM(alt_count),0) AS ref_freq,\n",
    "                                          SUM(ref_count)+SUM(alt_count) AS num_sample_alleles,\n",
    "                                          SUM(ref_count) AS ref_cnt,\n",
    "                                          SUM(alt_count) AS alt_cnt,\n",
    "                                          FROM cte2 GROUP BY 1,2,3,4\"\n",
    "\n",
    "allele_freqeuency<-dbGetQuery(gcp_account, allele_frequency_sql)\n",
    "head(allele_freqeuency)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb8f12a7-527b-4230-8389-625527cd7277",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Slideshow",
  "environment": {
   "kernel": "ir",
   "name": ".m115",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/:m115"
  },
  "kernelspec": {
   "display_name": "R (Local)",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
